File: typescript/src/flows/charge.test.ts
/**
 * Flow: create (mode=charge) → sign → charge
 *
 * Run:
 *   set -a && source ../../.env && set +a
 *   npm test -- --testNamePattern="charge"
 */

import {
  newClient, accountWallet, discoverPaymentMethod,
  createAndSign, signEip1559, pollUntilStatus, getEnv,
} from '../helpers.js'

test('charge (one-shot)', async () => {
  const client    = newClient()
  const accWallet = accountWallet()
  const pm        = await discoverPaymentMethod(client)
  const amount    = getEnv('AMOUNT', '1000000')

  console.log('→ creating payment (mode=charge) and submitting payer signature')
  const { paymentId } = await createAndSign(client, pm, 'charge', amount)
  console.log(`  payment_id=${paymentId}`)

  console.log('→ charge/payload')
  const prep = await client.payments.chargePayload(paymentId)
  expect(prep.unsignedTransaction).toBeTruthy()

  const signed = await signEip1559(prep.unsignedTransaction, accWallet)
  await client.payments.charge(paymentId, { signedTransaction: signed })

  const final = await pollUntilStatus(client, paymentId, ['charged'], 'charge')
  console.log(`  charged — status=${final.status}`)
})

File: go/flows/charge_test.go
package flows_test

// Flow: create (mode=charge) → sign → charge
//
// Run:
//   set -a && source ../../.env && set +a
//   go test ./flows/ -run TestCharge -v

import (
	"context"
	"testing"

	rail0 "github.com/rail0/go-sdk"
)

func TestCharge(t *testing.T) {
	client     := newClient(t)
	accountKey := loadKey(t, "ACCOUNT_PRIVATE_KEY")
	pm         := discoverPaymentMethod(t, client)

	t.Log("→ creating payment (mode=charge) and submitting payer signature")
	paymentID := createAndSign(t, client, pm, "charge")
	t.Logf("  payment_id=%s", paymentID)

	t.Log("→ charge/payload")
	prep, err := client.Payments.ChargePrepare(context.Background(), paymentID)
	if err != nil {
		t.Fatalf("ChargePrepare: %v", err)
	}
	signed := signEIP1559(t, prep.UnsignedTransaction, accountKey)
	if _, err := client.Payments.Charge(context.Background(), paymentID, rail0.SubmitTransactionRequest{
		SignedTransaction: signed,
	}); err != nil {
		t.Fatalf("Charge: %v", err)
	}
	final := pollUntilStatus(t, client, paymentID, "charged")
	t.Logf("  charged — status=%s", final.Status)
}
